src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/FallThroughCheck.java

```diff
@@ -102,17 +64,118 @@
  * To configure the check:
  * </p>
  * <pre>
- * &lt;module name=&quot;FallThrough&quot;/&gt;
+ * &lt;module name="FallThrough"/&gt;
+ * </pre>
+ * <p>
+ * Example:
+ * </p>
+ * <pre>
+ * switch (i) {
+ *   case 1:
+ *     i++;
+ *   case 2: // violation, case 1 contains code but lacks break, return, throw or continue statement
+ *     i++;
+ *     break;
+ *   case 3: // OK
+ *     i++;
+ *     return;
+ *   case 4: // OK
+ *     i++;
+ *     throw new Exception();
+ *   case 5: // OK
+ *     i++;
+ *     continue;
+ *   case 6: // OK
+ *   case 7: // OK, case 6 does not contain code
+ *     i++;
+ *   // OK, by default the last case is not checked
+ * }
+ * </pre>
+ * <p>
+ * Unreachable code still causes violation:
+ * </p>
+ * <pre>
+ * switch (i) {
+ *   default:
+ *     return;
+ *   case 1:
+ *     i++; // unreachable
+ *   case 2: // violation
+ *     return;
+ * }
  * </pre>
  * <p>
- * or
+ * Add special comments to suppress the warning:
+ * </p>
+ * <pre>
+ * switch (i) {
+ *   case 1:
+ *     i++; // fall through
+ *
+ *   case 2: // OK
+ *     i++;
+ *     // fallthru
+ *   case 3: { // OK
+ *     i++;
+ *   }
+ *   &#47;* fall-thru *&#47;
+ *   case 4: // OK
+ *   i++;
+ *     // Fallthru
+ *   case 5: // violation, "Fallthru" in case 4 should be "fallthru"
+ *     i++;
+ *     // fall through
+ *     i++;
+ *   case 6: // violation, the comment must be on the last non-empty line before 'case'
+ *     i++;
+ *   &#47;* fall through *&#47;case 7: // OK, comment can appear on the same line but before 'case'
+ *     i++;
+ * }
+ * </pre>
+ * <p>
+ * To configure the check so that it checks last case group
  * </p>
  * <pre>
  * &lt;module name=&quot;FallThrough&quot;&gt;
- *   &lt;property name=&quot;reliefPattern&quot; value=&quot;continue in next case&quot;/&gt;
+ *    &lt;property name=&quot;checkLastCaseGroup&quot; value=&quot;true&quot;/&gt;
  * &lt;/module&gt;
  * </pre>
  * <p>
+ * Example:
+ * </p>
+ * <pre>
+ * switch (i) {
+ *   case 1:
+ *     break;
+ *   case 2: // OK
```

Comment:
why ok if you have violation comment 2 lines below?

Reply:
I write `OK` or `violation` after each case because that's where the error appears. So `OK` in case 2 is actually from case 1.
I just realized it's kind of hard to understand. Do you want me to change that?

Reply:
It is best to write violation comment on line where it is reported, same for ok comments (where violation would be if it wasn't ok)

Reply:
Ok. So the way I'm writing right now follows that rule.

Reply:
from CLI in PR:
```
cat Test.java
public class Test {
  public void foo() {
    int i = 0;
    switch (i) {
      case 1:
        break;
      case 2: // OK // line 7 // violation comment should be there!!!! 
        i++;
        // violation, last case group is checked
    }
  }
}
java -jar .\checkstyle-8.36.1-all.jar -c config.xml Test.java
Starting audit...
[ERROR] E:\checkstyle-playground\Test.java:7:7: Fall through f
```

Reply:
@anhminhtran235, please address this

Reply:
@romani So 'OK' and 'violation' will be on the same line?
Can I write like this?
```
public class Test {
  public void foo() {
    int i = 0;
    switch (i) {
      case 1:
        break;
      case 2: // Previous case: OK 
              // This case: violation, last case must have statement that transfer control
        i++;
    }
  }
}
```

Reply:
`violation` MUST be on same line a it CLI report it (there are some exceptions from this rule, but this case is not exception).
`ok` should be on line where violation might occur.
```
$ cat config.xml 
<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE module PUBLIC
  "-//Puppy Crawl//DTD Check Configuration 1.3//EN"
  "http://www.puppycrawl.com/dtds/configuration_1_3.dtd">

<module name="Checker">
    <module name="TreeWalker">
           <module name="FallThrough">
             <property name="checkLastCaseGroup" value="true"/>
           </module>
    </module>
</module>

$ cat Test.java 
public class Test {
  public void foo() {
    int i = 0;
    switch (i) {
      case 1: // OK
        break;
      case 2: // violation, last case must have statement that transfer control
        i++;
    }
  }
}

$ java -jar checkstyle-8.36.1-all.jar -c config.xml Test.java
Starting audit...
[ERROR] /var/tmp/Test.java:7:7: Fall through from the last branch of the switch statement. [FallThrough]
Audit done.
Checkstyle ends with 1 errors.
```

Reply:
@romani  `violation` occurs at line 7 `twice`. 1 for the previous case and 1 for the last case. That means if `violation` occurs, it occurs at line 7. So should `OK` be on line 7 too?
Sorry for the trouble. I just want to make sure I understand this one.
```
cat config.xml
<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE module PUBLIC
  "-//Puppy Crawl//DTD Check Configuration 1.3//EN"
  "http://www.puppycrawl.com/dtds/configuration_1_3.dtd">

<module name="Checker">
    <module name="TreeWalker">
           <module name="FallThrough">
             <property name="checkLastCaseGroup" value="true"/>
           </module>
    </module>
</module>

cat Test.java
public class Test {
  public void foo() {
    int i = 0;
    switch (i) {
      case 1:
        i++;
      case 2: // violation happens at this line TWICE!
        i++;
    }
  }
}

java -jar .\checkstyle-8.36.1-all.jar -c config.xml Test.java
Starting audit...
[ERROR] E:\checkstyle-playground\Test.java:7:7: Fall through from previous branch of the switch statement. [FallThrough]
[ERROR] E:\checkstyle-playground\Test.java:7:7: Fall through from the last branch of the switch statement. [FallThrough]
Audit done.
Checkstyle ends with 2 errors.
```

Reply:
` // violations. ... Some words...`

Reply:
No. What I mean is because case 1 and case 2's violations will occur at the same line, so I should write like this
```
public class Test {
  public void foo() {
    int i = 0;
    switch (i) {
      case 1: // <- OK should not be here because this is not where the error is
        break;
      case 2: // Previous case: OK <- OK should be here
              // This case: violation, last case must have statement that transfer control
        i++;
    }
  }
}
```
Do you see what I mean?

Reply:
Please share exact config for this java file and CLI output. Violation comment should be on same line number as CLI report.

Reply:
It's [right above](https://github.com/checkstyle/checkstyle/pull/8847#discussion_r496970680). I substitute `break `with `i++` to generate an error.

Reply:
And I already provided expected comment :)

Reply:
You mean [this reply](https://github.com/checkstyle/checkstyle/pull/8847#discussion_r496752290)? 
You said '`ok` should be on line where violation might occur.' But in the code you share, it's not:
```
$ cat Test.java 
public class Test {
  public void foo() {
    int i = 0;
    switch (i) {
      case 1: // OK <- I don't know why OK should be here. If there is any violation, it does not occur here.
                 // The program needs to reach `case 2` to know something went wrong.
        break;
      case 2: // violation, last case must have statement that transfer control
        i++;
    }
  }
}
```

Reply:
```
switch (i) {
      case 1: // Case 1 has Fall through violation
        i++;
      case 2: // But it is reported at this line. So `OK` and `violation` should be here
        ...
    }
```

Reply:
if you showed work as following, we would resolve issue much quicker:
```
$ cat config.xml 
<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE module PUBLIC
  "-//Puppy Crawl//DTD Check Configuration 1.3//EN"
  "http://www.puppycrawl.com/dtds/configuration_1_3.dtd">

<module name="Checker">
    <module name="TreeWalker">
           <module name="FallThrough">
             <property name="checkLastCaseGroup" value="true"/>
           </module>
    </module>
</module>
 
$ cat -n Test.java 
     1  public class Test {
     2    public void foo() {
     3      int i = 0;
     4      switch (i) {
     5        case 1: // OK <- I don't know why OK should be here. 
     6                   // The program needs to reach `case 2` to know something went wrong.
     7          break;
     8        case 2: // violation, last case must have statement that transfer control
     9          i++;
    10      }
    11    }
    12  }

$ java -jar checkstyle-8.36.1-all.jar -c config.xml Test.java
Starting audit...
[ERROR] /var/tmp/Test.java:8:7: Fall through from the last branch of the switch statement. [FallThrough]
Audit done.
Checkstyle ends with 1 errors.

$ cat -n Test.java 
     1  public class Test {
     2    public void foo() {
     3      int i = 0;
     4      switch (i) {
     5        case 1: // OK <- I don't know why OK should be here. 
     6                   // The program needs to reach `case 2` to know something went wrong.
     7          i--;
     8        case 2: // violation, last case must have statement that transfer control
     9          i++;
    10      }
    11    }
    12  }

$ java -jar checkstyle-8.36.1-all.jar -c config.xml Test.java
Starting audit...
[ERROR] /var/tmp/Test.java:8:7: Fall through from previous branch of the switch statement. [FallThrough]
[ERROR] /var/tmp/Test.java:8:7: Fall through from the last branch of the switch statement. [FallThrough]
Audit done.
Checkstyle ends with 2 errors.
```

Yes, "//ok" should not be on case that we discuss in this review thread.
But please fix example in code, and place `violation` on line with `case` , not a separate line.

Please fix code and please regenerate html website, lets do final review.

Reply:
@romani Code fixed and images fixed :)